docs/tutorials/dense-subgraph-gbs.ipynb

```diff
@@ -0,0 +1,430 @@
+      "62. density=0.7142857142857143\n",
+      "63. density=0.7142857142857143\n",
+      "64. density=0.7142857142857143\n",
+      "65. density=0.7142857142857143\n",
+      "66. density=0.7142857142857143\n",
+      "67. density=0.7142857142857143\n",
+      "68. density=0.7142857142857143\n",
+      "69. density=0.7142857142857143\n",
+      "70. density=0.7142857142857143\n",
+      "71. density=0.7142857142857143\n",
+      "72. density=0.7142857142857143\n",
+      "73. density=0.7142857142857143\n",
+      "74. density=0.7142857142857143\n",
+      "75. density=0.7142857142857143\n",
+      "76. density=0.7857142857142857\n",
+      "77. density=0.7857142857142857\n",
+      "78. density=0.7857142857142857\n",
+      "79. density=0.7857142857142857\n",
+      "80. density=0.7857142857142857\n",
+      "81. density=0.7857142857142857\n",
+      "82. density=0.7857142857142857\n",
+      "83. density=0.7857142857142857\n",
+      "84. density=0.7857142857142857\n",
+      "85. density=0.7857142857142857\n",
+      "86. density=0.7857142857142857\n",
+      "87. density=0.7857142857142857\n",
+      "88. density=0.7857142857142857\n",
+      "89. density=0.7857142857142857\n",
+      "90. density=0.7857142857142857\n",
+      "91. density=0.7857142857142857\n",
+      "92. density=0.7857142857142857\n",
+      "93. density=0.7857142857142857\n",
+      "94. density=0.7857142857142857\n",
+      "95. density=0.7857142857142857\n",
+      "96. density=0.7857142857142857\n",
+      "97. density=0.7857142857142857\n",
+      "98. density=0.7857142857142857\n",
+      "99. density=0.7857142857142857\n",
+      "Solution: Graph with 8 nodes and 22 edges\n",
+      "Density: 0.7857142857142857\n"
+     ]
+    }
+   ],
+   "source": [
+    "initial_temperature = 100  # Initial temperature for simulated annealing\n",
+    "alpha = 0.85  # Parameter for geometric cooling\n",
+    "ITER = 100  # Number of iterations\n",
+    "\n",
+    "subgraph = generate_dense_subgraph(GRAPH, k)\n",
+    "\n",
+    "best_subgraph_so_far = subgraph\n",
+    "best_density_so_far = nx.density(subgraph)\n",
+    "\n",
+    "T = initial_temperature\n",
+    "\n",
+    "for i in range(ITER):\n",
+    "    density = nx.density(subgraph)\n",
+    "    print(f\"{i}. density={density}\")\n",
+    "    new_subgraph = gbs_tweak(subgraph, l)\n",
+    "    new_density = nx.density(new_subgraph)\n",
+    "\n",
+    "    if new_density > density:\n",
+    "        subgraph = new_subgraph\n",
+    "    else:\n",
+    "        probability = np.exp((new_density - density) / T)\n",
+    "\n",
+    "        if np.random.rand() < probability:\n",
+    "            # Accept anyway with some probability\n",
+    "            subgraph = new_subgraph\n",
+    "\n",
+    "    if new_density > best_density_so_far:\n",
+    "        best_subgraph_so_far = new_subgraph\n",
+    "        best_density_so_far = new_density\n",
+    "\n",
+    "    T *= alpha  # Decrease temperature\n",
```

Comment:
I think briefly explaining this logic and referencing simulated annealing could help the reader.